src/datafusion/helpers.rs

```diff
@@ -0,0 +1,228 @@
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::Name(_) => {
+                internal_err!("Function `Expr` with name should be resolved.")
+            }
+        },
+
+        Expr::AggregateFunction { .. }
+        | Expr::Sort { .. }
+        | Expr::WindowFunction { .. }
+        | Expr::Wildcard { .. }
+        | Expr::Unnest { .. }
+        | Expr::Placeholder(_) => {
+            is_applicable = false;
+            Ok(VisitRecursion::Stop)
+        }
+    })
+    .unwrap();
+    is_applicable
+}
+
+// Below is all the logic necessary (I think) to convert a PhysicalExpr into a ZarrChunkFilter.
+// The logic is mostly copied from datafusion, and is simplified here for the zarr use case.
+pub struct ZarrFilterCandidate {
+    expr: Arc<dyn PhysicalExpr>,
+    projection: Vec<usize>,
+}
+
+struct ZarrFilterCandidateBuilder<'a> {
+    expr: Arc<dyn PhysicalExpr>,
+    file_schema: &'a Schema,
+    required_column_indices: BTreeSet<usize>,
+}
+
+impl<'a> ZarrFilterCandidateBuilder<'a> {
+    pub fn new(expr: Arc<dyn PhysicalExpr>, file_schema: &'a Schema) -> Self {
+        Self {
+            expr,
+            file_schema,
+            required_column_indices: BTreeSet::default(),
+        }
+    }
+
+    pub fn build(mut self) -> DataFusionResult<Option<ZarrFilterCandidate>> {
+        let expr = self.expr.clone().rewrite(&mut self)?;
+
+        Ok(Some(ZarrFilterCandidate {
+            expr,
+            projection: self.required_column_indices.into_iter().collect(),
+        }))
+    }
+}
+
+impl<'a> TreeNodeRewriter for ZarrFilterCandidateBuilder<'a> {
+    type N = Arc<dyn PhysicalExpr>;
+
+    fn pre_visit(&mut self, node: &Arc<dyn PhysicalExpr>) -> DataFusionResult<RewriteRecursion> {
+        if let Some(column) = node.as_any().downcast_ref::<Column>() {
+            if let Ok(idx) = self.file_schema.index_of(column.name()) {
+                self.required_column_indices.insert(idx);
+            }
+        }
+
+        Ok(RewriteRecursion::Continue)
+    }
+
+    fn mutate(&mut self, expr: Arc<dyn PhysicalExpr>) -> DataFusionResult<Arc<dyn PhysicalExpr>> {
+        Ok(expr)
+    }
+}
+
+#[derive(Clone)]
+pub struct ZarrDatafusionArrowPredicate {
+    physical_expr: Arc<dyn PhysicalExpr>,
+    projection_mask: ZarrProjection,
+    projection: Vec<String>,
+}
+
+impl ZarrDatafusionArrowPredicate {
+    pub fn new(candidate: ZarrFilterCandidate, schema: &Schema) -> DataFusionResult<Self> {
+        let cols: Vec<_> = candidate
+            .projection
+            .iter()
+            .map(|idx| schema.field(*idx).name().to_string())
+            .collect();
+
+        let schema = Arc::new(schema.project(&candidate.projection)?);
+        let physical_expr = reassign_predicate_columns(candidate.expr, &schema, true)?;
```

Comment:
If I understand correctly, the physical expression has the name of each column as well as an index for each. Since it was first created off of a `Expr`, using the file schema, the indices for each column don't necessarily match what they will be in the record batch we pass to the physical expression. Assuming we will pass the physical expression a record batch that only contains the columns it needs, we need to remap indices to columns, e.g. we go from `(lat, 1), (lon, 2)` to `(lat, 0), (lon, 1)`.